xla/service/gpu/transforms/memory_space_propagation.cc

```diff
@@ -0,0 +1,95 @@
+/* Copyright 2025 The OpenXLA Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "xla/service/gpu/transforms/memory_space_propagation.h"
+
+#include <optional>
+#include <utility>
+
+#include "absl/container/flat_hash_set.h"
+#include "absl/log/check.h"
+#include "absl/status/statusor.h"
+#include "absl/strings/string_view.h"
+#include "xla/hlo/analysis/hlo_dataflow_analysis.h"
+#include "xla/hlo/ir/hlo_instruction.h"
+#include "xla/hlo/ir/hlo_opcode.h"
+#include "xla/layout.h"
+#include "xla/layout_util.h"
+#include "xla/service/hlo_value.h"
+#include "xla/shape.h"
+#include "xla/shape_util.h"
+
+namespace xla::gpu {
+namespace {
+
+bool Propagate(ShapeIndexView index, HloInstruction* instruction,
+               const Shape& src_shape, const HloDataflowAnalysis* analysis) {
+  bool modified = false;
+
+  const HloValue& value =
+      analysis->GetValueDefinedAt(instruction, ShapeIndex(index));
+  auto memory_space_to_propagate = src_shape.layout().memory_space();
+  for (const HloPosition& position : value.positions()) {
+    auto* instruction = position.instruction;
+    Shape* shape = ShapeUtil::GetMutableSubshape(instruction->mutable_shape(),
+                                                 position.index);
+    if (!shape->has_layout() ||
+        shape->layout().memory_space() == memory_space_to_propagate) {
+      continue;
+    }
+    shape->mutable_layout()->set_memory_space(memory_space_to_propagate);
+    modified = true;
+  }
+
+  // We don't propagate from S(x) operand to S(x) output because it's
+  // pure memory transfer without any computation in S(x). Computations
+  // including transposes should be defined explicitly in JAX.
```

Comment:
Can you explain this comment? I don't understand how they are relevant to the code here. In this routine, we propagate the memory space from the definition of the value to the use of the value, right?

Reply:
I replied in another comment. I just hope this memory space propagation be simple and not placing any unexpected behavior for the users. When user annotate a buffer to be in some memory space, its definition and all references should be annotated but nothing beyond.